First draft of JSON Schema

scottdchristian opened the pull request:
The users have requested defined parameters for the data they will be
submitting.  This document has the goal of providing such guidance.  I
placed the schemas in a folder in the db folder.  However it should go
in ""https://e-manifest.18f.gov/schemas/manifestSchema20160105.json"

scottdchristian committed:
First draft of JSON Schema

The users have requested defined parameters for the data they will be
submitting.  This document has the goal of providing such guidance.  I
placed the schemas in a folder in the db folder.  However it should go
in ""https://e-manifest.18f.gov/schemas/manifestSchema20160105.json"

jessieay commented:
why is there a date on this schema? seems like we could just call it `manifestSchema` and update its content as required?

scottdchristian commented:
Thanks, actually force of bad habit.

A major point of GitHub is version control.  You are right I should have just named it how we would use it rather than my desktop version control.  I will rename and edit in the morning then refresh.

Thanks again,
Scott

From: Jessie A. Young [mailto:notifications@github.com]
Sent: Tuesday, January 05, 2016 4:59 PM
To: 18F/e-manifest e-manifest@noreply.github.com
Cc: Christian, Scott Christian.Scott@epa.gov
Subject: Re: [e-manifest] First draft of JSON Schema (#13)

why is there a date on this schema? seems like we could just call it manifestSchema and update its content as required?

—
Reply to this email directly or view it on GitHubhttps://github.com/18F/e-manifest/pull/13#issuecomment-169146825.

jeremiak commented on db/schemas/manifestSchema20160105.json:
```diff
@@ -0,0 +1,367 @@
+            }
+        },
+
+        "TelephoneNumberDataType":  {
```
Looks like we're using snake case everywhere else so this field should be `telephone_number_data_type`

jessieay replied:
I've always used snake case when creating JSON APIs with Ruby because that is a Ruby convention. 

I recently read, however, that it is preferred to use snake case since JSON is JavaScript and js convention is to use CamelCase. 

I am fine with either as long as we are consistent.

pkarman replied:
+1 for consistency.

It's true that Google and others consistently use camelCase for JSON. schema.org does that as well.

More tellingly, since this schema purports to adhere to json-schema.org conventions, the examples there all camelCase.

pkarman replied:
having said all that, I don't think it really matters to consumers of the API, as long as it is consistent and well-documented.

scottdchristian replied:
I will go with consistency.  For this project the initial data was written in snake_case, since users have already started integrating the JSON and I do not see a compelling reason to change the field names now.  

With that said, my next iteration (after this one) is to add new fields to the Schema.

jeremiak commented on db/schemas/manifestSchema20160105.json:
```diff
@@ -0,0 +1,367 @@
+                    "description": "Address line 1",
+                    "type": "string",
+                    "maxLength":30 ,
+                    "minLength":0 
```
Should these actually be something like `"minLength": 1` so that we get a validation error if these values are not supplied?

scottdchristian committed:
Merge remote-tracking branch 'refs/remotes/18F/master' into Adding-JSON-Schema-First-Draft

scottdchristian committed:
Changes made per comments made on the first draft.

"Second draft of JSON Schema.  This version, changed file name, enforced
snake_case, and removed minLengths=0. JSON Schema uses camelCase so for
keywords, I left those in camelCase and everything else in snake_case.

jessieay commented on db/schemas/manifestschema.json:
```diff
@@ -0,0 +1,356 @@
+            }
+        },
+
+        "telephonenumberdatatype":  {
```
should be `telephone_number_data_type`

scottdchristian replied:
OK, but we need a plan to make this change.  The name included is the name of the field in industry.html and the one posted in our sample.  Let's discuss at story time.  FWIW, I am not opposed to changing the names much.  We'll be adding new fields anyway.

jessieay commented on db/schemas/manifestschema.json:
```diff
@@ -0,0 +1,356 @@
+                    }
+                    }
+                  },
+
```
let's remove these extra newlines

scottdchristian replied:
Newbie question, which lines to be specific?

jessieay replied:
270 and 271 :)

scottdchristian replied:
Thanks

jessieay commented on db/schemas/manifestschema.json:
```diff
@@ -0,0 +1,356 @@
+
+                "manifest_items": {
+                    "type": "array",
+                    "items": [{
```
I think the opening curly brace on this line (and its closing curly brace) are not needed since this is an array. Or is it array of objects and there is one object?

pkarman replied:
I would recommend we just run the whole file through a JSON pretty printer so there's no question of manual formatting (and makes diffs better later on).

jessieay replied:
:+1:

scottdchristian replied:
Looking at the JSON and form it is an array of one object that get's repeated.  manifest_items is a section of the form and all those pieces are part of that section.  A manifest can have one manifest item or many manifest items. On the form items 9a-13 (http://www3.epa.gov/epawaste/hazard/transportation/manifest/pdf/newform.pdf)

scottdchristian replied:
I didn't answer the question directly:  Yes its an array of one object

jessieay commented:
Going to merge as-is so I can start working on validating that our JSON matches this schema
